File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedPlaySound.java
package net.easecation.ghosty.recording.level.updated;

import cn.nukkit.level.Level;
import cn.nukkit.math.BlockVector3;
import cn.nukkit.network.protocol.PlaySoundPacket;
import cn.nukkit.utils.BinaryStream;
import net.easecation.ghosty.recording.level.LevelRecordNode;

public class LevelUpdatedPlaySound implements LevelUpdated {

    public String name;
    public int x;
    public int y;
    public int z;
    public float volume;
    public float pitch;

    private LevelUpdatedPlaySound(String name, int x, int y, int z, float volume, float pitch) {
        this.name = name;
        this.x = x;
        this.y = y;
        this.z = z;
        this.volume = volume;
        this.pitch = pitch;
    }

    public static LevelUpdatedPlaySound of(PlaySoundPacket packet) {
        return new LevelUpdatedPlaySound(packet.name, packet.x, packet.y, packet.z, packet.volume, packet.pitch);
    }

    public LevelUpdatedPlaySound(BinaryStream stream) {
        this.read(stream);
    }

    @Override
    public int getUpdateTypeId() {
        return LevelUpdated.TYPE_PLAY_SOUND;
    }

    @Override
    public void processTo(LevelRecordNode node) {
        PlaySoundPacket packet = new PlaySoundPacket();
        packet.name = this.name;
        packet.x = this.x;
        packet.y = this.y;
        packet.z = this.z;
        packet.volume = this.volume;
        packet.pitch = this.pitch;
        node.handleLevelChunkPacket(Level.chunkHash(this.x >> 4, this.z >> 4), packet);
    }

    @Override
    public void write(BinaryStream stream) {
        stream.putString(name);
        stream.putBlockVector3(new BlockVector3(x, y, z));
        stream.putLFloat(volume);
        stream.putLFloat(pitch);
    }

    @Override
    public void read(BinaryStream stream) {
        this.name = stream.getString();
        BlockVector3 pos = stream.getBlockVector3();
        this.x = pos.x;
        this.y = pos.y;
        this.z = pos.z;
        this.volume = stream.getLFloat();
        this.pitch = stream.getLFloat();
    }

    @Override
    public String toString() {
        return "LevelUpdatedPlaySound{" +
            "name='" + name + '\'' +
            ", x=" + x +
            ", y=" + y +
            ", z=" + z +
            ", volume=" + volume +
            ", pitch=" + pitch +
            '}';
    }
}

File: src/main/java/net/easecation/ghosty/recording/level/updated/LevelUpdatedPlaySound.kt
package net.easecation.ghosty.recording.level.updated

import cn.nukkit.level.Level
import cn.nukkit.math.BlockVector3
import cn.nukkit.network.protocol.PlaySoundPacket
import cn.nukkit.utils.BinaryStream
import kotlinx.serialization.Serializable
import net.easecation.ghosty.recording.level.LevelRecordNode

@Serializable
data class LevelUpdatedPlaySound(
    val name: String,
    val x: Int,
    val y: Int,
    val z: Int,
    val volume: Float,
    val pitch: Float
) : LevelUpdated {

    override fun getUpdateTypeId(): Int = LevelUpdated.TYPE_PLAY_SOUND

    override fun processTo(node: LevelRecordNode) {
        PlaySoundPacket().also {
            it.name = name
            it.x = x
            it.y = y
            it.z = z
            it.volume = volume
            it.pitch = pitch
        }.also { packet ->
            node.handleLevelChunkPacket(Level.chunkHash(x shr 4, z shr 4), packet)
        }
    }

    override fun backwardTo(node: LevelRecordNode) = Unit

    companion object {
        @JvmStatic
        fun of(packet: PlaySoundPacket) =
            LevelUpdatedPlaySound(packet.name, packet.x, packet.y, packet.z, packet.volume, packet.pitch)

        @JvmStatic
        val ADAPTER: LevelUpdateAdapter<LevelUpdatedPlaySound> = Adapter
    }

    private object Adapter : LevelUpdateAdapter<LevelUpdatedPlaySound> {
        override fun write(updated: LevelUpdatedPlaySound, stream: BinaryStream) {
            stream.putString(updated.name)
            stream.putBlockVector3(BlockVector3(updated.x, updated.y, updated.z))
            stream.putLFloat(updated.volume)
            stream.putLFloat(updated.pitch)
        }

        override fun read(stream: BinaryStream, formatVersion: Int): LevelUpdatedPlaySound {
            val name = stream.string
            val pos = stream.blockVector3
            val volume = stream.lFloat
            val pitch = stream.lFloat
            return LevelUpdatedPlaySound(name, pos.x, pos.y, pos.z, volume, pitch)
        }
    }
}